### ISSUE TITLE
Sessions

### ISSUE THREAD

**Issue by xcesaralejandro** [xcesaralejandro/lti1p3#7]
When integrating an LTI within an LMS, it is common for it to load in an iframe, which can cause issues due to new cookie policies in browsers. Until recently, it was sufficient to set same_site as none and secure as true to maintain sessions. However, with the introduction of CHIPS (Cookies Having Independent Partitioned State), this approach is no longer enough.

In Laravel 11, there is an environment key called SESSION_PARTITIONED_COOKIE that enables CHIPS. You can find more information about this technology in the official Google documentation.

Although some solutions exist to address the cookie issue, a definitive solution has yet to be adopted. IMS Global has attempted to establish a standard based on post messaging, but this is not supported by all LMS. On the other hand, CHIPS simplifies cookie management without additional configurations, although it is still not implemented in browsers like Safari (as of this writing). It is likely that CHIPS will become the primary solution, relegating the IMS Global standard to a secondary role.

Currently, due to these changes, sessions started with the package in the onResourceLinkRequest and onDeepLinkingRequest methods are lost, as cookies are not set until a final redirection is made from one of these methods. To start the session correctly, you must redirect to a route within your application, under your domain, so the cookie is properly set.

Initially, adding an additional redirection point to start the session and then returning to the original launch methods in the package could be considered. However, this would add unnecessary complexity to the code. As a developer, you can redirect directly to a new entry point in your application and manually authenticate the user, avoiding additional redirects.

Example:

````
public function onResourceLinkRequest(string $instance_id): mixed {
    return redirect()->route('build-session', ['instanceId' => $instance_id]);
}
````

Current user authentication:

````
Route::middleware('web')->group(function () {
    Route::get('/build-session/{instanceId}', function ($instanceId) {
        $instance = LtiInstance::find($instanceId);
        Auth::login($instance->user, true);
        return view('example');
    })->name('build-session');
});
````

Previously, we could perform the authentication directly within onResourceLinkRequest or onDeepLinkingRequest with:

````
$instance = LtiInstance::find($instanceId);
Auth::login($instance->user, true);
````

Now, this is no longer possible, as at that point, the cookie is not set correctly. Therefore, it is necessary to redirect to another route in your application, passing the available LTI instance and authenticating the user there. Once this authentication is done, your application will function normally.

**Comment by felimoles** [comment]
maybe storage api solve this https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API

**Comment by xcesaralejandro** [comment]
Yes, it’s an alternative, but some browsers require user confirmation to use the Storage Access API, unlike CHIPS and postMessage(). Relying on user acceptance can lead to additional issues if they accidentally reject cookies. Just as there are extensions to block ads, some students might have extensions that automatically deny cookie consent.

Canvas LMS has already implemented postMessage, but they even note that it’s not compatible with their mobile app, which limits our options. There are alternatives, but none seem to provide a clear advantage; each one follows its own standard, ultimately making things more difficult for developers.

The most reliable option would be to adapt the lti instance generated by the package, but this complicates the code and slows down the development of an LTI built only with Laravel (assuming security is a priority). The best solution would be to build a frontend with Vue, React, or Angular, using Laravel only to provide an API, with the lti instance serving as an authentication/authorization token.

Canvas LMS Ref:
https://canvas.instructure.com/doc/api/file.lti_window_post_message.html
https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage
